module/zfs/dsl_scan.c

```diff
@@ -3567,6 +3864,447 @@ dsl_process_async_destroys(dsl_pool_t *dp, dmu_tx_t *tx)
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg &&
+	    snap_obj_txg <= txg_to_consider) {
+
+		error = dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds);
+		if (error != 0)
+			goto out;
+
+		if (dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj != fs) {
+			snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+			snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+			dsl_dataset_rele(ds, FTAG);
+			continue;
+		}
+
+		boolean_t affected = B_TRUE;
+		if (check_snapshot) {
+			uint64_t blk_txg;
+			error = find_birth_txg(ds, zep, &blk_txg);
+
+			/*
+			 * Scrub the snapshot also when zb_birth == 0 or when
+			 * find_birth_txg() returns an error.
+			 */
+			affected = (error == 0 && zep->zb_birth == blk_txg) ||
+			    (error != 0) || (zep->zb_birth == 0);
+		}
+
+		/* Scrub snapshots. */
+		if (affected) {
+			snap_obj_array[aff_snap_count] = snap_obj;
+			aff_snap_count++;
+
+			zbookmark_phys_t zb;
+			zep_to_zb(snap_obj, zep, &zb);
+			scn->scn_zio_root = zio_root(spa, NULL, NULL,
+			    ZIO_FLAG_CANFAIL);
+			/* We have already acquired the config lock for spa */
+			read_by_block_level(scn, zb);
+
+			(void) zio_wait(scn->scn_zio_root);
+			scn->scn_zio_root = NULL;
+
+			scn->errorscrub_phys.dep_examined++;
+			scn->errorscrub_phys.dep_to_examine--;
+			(*count)++;
+			if ((*count) == zfs_scrub_error_blocks_in_one_txg ||
+			    dsl_error_scrub_check_suspend(scn, &zb)) {
+				dsl_dataset_rele(ds, FTAG);
+				error = SET_ERROR(EFAULT);
+				goto out;
+			}
+		}
+		snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+		snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+		dsl_dataset_rele(ds, FTAG);
+	}
+
+	if (zap_clone != 0 && aff_snap_count > 0) {
+		error = scrub_clones(spa, zap_clone, snap_count, snap_obj_array,
+		    zep, count);
+	}
+
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (error);
+}
+
+/*
+ * Clone scrubbing. Modelled after check_clones().
+ */
+static int scrub_clones(spa_t *spa, uint64_t zap_clone, uint64_t snap_count,
+    uint64_t *snap_obj_array, zbookmark_err_phys_t *zep, int *count)
+{
+	int error = 0;
+	zap_cursor_t *zc;
+	zap_attribute_t *za;
+
+	zc = kmem_zalloc(sizeof (zap_cursor_t), KM_SLEEP);
+	za = kmem_zalloc(sizeof (zap_attribute_t), KM_SLEEP);
+
+	for (zap_cursor_init(zc, spa->spa_meta_objset, zap_clone);
+	    zap_cursor_retrieve(zc, za) == 0;
+	    zap_cursor_advance(zc)) {
+
+		dsl_pool_t *dp = spa->spa_dsl_pool;
+		dsl_dataset_t *clone;
+		error = dsl_dataset_hold_obj(dp, za->za_first_integer,
+		    FTAG, &clone);
+
+		if (error != 0)
+			break;
+
+		/*
+		 * Only clones whose origins were affected could also
+		 * have affected snapshots.
+		 */
+		boolean_t found = B_FALSE;
+		for (int i = 0; i < snap_count; i++) {
+			if (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+			    == snap_obj_array[i])
+				found = B_TRUE;
+		}
+		dsl_dataset_rele(clone, FTAG);
+
+		if (!found)
+			continue;
+
+		error = scrub_filesystem(spa, za->za_first_integer, zep,
```

Comment:
> For nested datasets there is a zfs_max_dataset_nesting limit (default 50) to keep things from getting too deep. However, I don't see any limit for clone+snapshot recursion, perhaps we should introduce a zfs_max_close_nesting limit to at least bound this case. As long as we can keep the stack frame size here to say <100 bytes, then it seems like a clone depth of 50 here should be workable (~5k out of a 16k stack). Alternately, we could check the stack depth and return a unique error when we're within say a page of the limit.

@behlendorf I quoted your comment here because it went missing after some pushes. Do you have any tips in calculating stack depth or size of a function?

I tried `offset of first allocated variable` - `offset of last allocated variable` but a) I am not sure if this is going to work in all archs and b) I get vastly different results from `checkstack.pl` in the linux kernel.

Reply:
It would be a good idea to do the same in `check_filesystem()`. 

In my ordinary VMs (PAGE_SIZE 4k, kernel stack size 8k) I can go in a depth of 43 filesystems before `check_filesystem()` (zpool status -v) or `scrub_filesystem()` (zpool scrub -e) crash.

Reply:
I think there may be an easier way around this: set scrub_filesystem() to suspend when it reaches a definite count of recursion (just like with `zfs_scrub_error_blocks_in_one_txg`).

Reply:
While trying to work around this limitation, I keep asking myself whether we *need* to scrub all clones. Isn't scrubbing the snapshots sufficient?
After all, I can see only two possibilities: A clone of a snapshot may:
a) have modified the error block in question (e.g. freeing or overwriting it)
b) have not touched it

In case (a) we do not have to scrub the clone, ie the error block was freed/overwritten.
In case (b) we have already scrubbed the origin snapshot, why scrub it again?

@behlendorf @ahrens am I missing something?

Reply:
I removed `scrub_clones()` and the iteration it introduced entirely.

Reply:
I believe that should be fine for exactly the reasons you mentioned.